File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> st;
         std::unordered_map<char, char> map = {
            {'(',')'},
            {'[',']'},
            {'{','}'}
        };

        for(auto ch: s){
            if(map.count(ch)){
                st.push(ch);
            }else{

                if(!st.empty() && map[st.top()] == ch){
                   st.pop();
                }else{
                   return false;
                }
            }
        }

        return st.empty();
        
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack =[]

        map = {
            "(":")",
            "[":"]",
            "{":"}"
        }


        for o in s:

            if o in map:
                stack.append(o)
            else:

                if not stack:
                    return False
                
                lastData = stack.pop()
                if map[lastData] != o:
                    return False
        return len(stack) == 0
        
        